include/seqan3/alphabet/union_alphabet.hpp

```diff
@@ -31,237 +31,411 @@
-struct union_alphabet
+class union_alphabet
 {
-    /* types */
 public:
-    //! The size of the alphabet, i.e. the number of different values it can take.
+    //!\brief The size of the alphabet, i.e. the number of different values it can take.
     static constexpr size_t value_size = (alphabet_types::value_size + ... + first_alphabet_type::value_size);
-    //! the type of the alphabet when converted to char (e.g. via @link to_char @endlink)
-    using char_type = typename first_alphabet_type::char_type;
-    //! the type of the alphabet when represented as a number (e.g. via @link to_rank @endlink)
-    using rank_type = decltype(detail::at_least_type<value_size>());
 
-protected:
-    using tuple_t = std::tuple<first_alphabet_type, alphabet_types...>;
-    using index_sequence_t = decltype(std::make_index_sequence<std::tuple_size_v<tuple_t>>{});
-    static constexpr auto sizes = detail::alphabet_sizes<tuple_t>(index_sequence_t{});
-    static constexpr auto alphabets = detail::alphabet_objects<tuple_t>(index_sequence_t{});
+    //!\brief The type of the alphabet when converted to char (e.g. via \link to_char \endlink)
+    using char_type = typename first_alphabet_type::char_type;
 
-public:
-    //! internal value
-    rank_type value;
+    //!\brief The type of the alphabet when represented as a number (e.g. via \link to_rank \endlink)
+    using rank_type = detail::min_viable_uint_t<value_size>;
+
+    /*!\brief The type used to assign a value from one of the base alphabets
+     * during copy construction or copy assignment.
+     *
+     * For example used by
+     * union_alphabet::union_alphabet(const variant_type & alphabet) or
+     * union_alphabet::operator=(const variant_type & alphabet)
+     */
+    using variant_type = std::variant<first_alphabet_type, alphabet_types...>;
+
+    //!\name Default constructors
+    //!\{
+    constexpr union_alphabet() = default;
+    constexpr union_alphabet(union_alphabet const &) = default;
+    constexpr union_alphabet(union_alphabet &&) = default;
+    //!\}
+
+    //!\name Default assignment operators
+    //!\{
+    constexpr union_alphabet & operator= (union_alphabet const &) = default;
+    constexpr union_alphabet & operator= (union_alphabet &&) = default;
+    //!\}
+
+    //!\name Conversion constructors
+    //\{
+    /*!\brief Construction via a value of the base alphabets
+     *
+     * ```cpp
+     *     union_alphabet<dna4, gap> letter1{dna4::C}; // or
+     *     union_alphabet<dna4, gap> letter2 = gap::GAP;
+     * ```
+     */
+    constexpr union_alphabet(variant_type const & alphabet) :
+        _value{from_base_(alphabet)}
+    {}
+    //!\}
+
+    //!\name Conversion assignment operators
+    //!\{
+    /*!\brief Assignment via a value of the base alphabets
+     *
+     * ```cpp
+     *     union_alphabet<dna4, gap> letter1{};
+     *     letter1 = gap::GAP;
+     * ```
+     */
+    constexpr union_alphabet & operator= (variant_type const & alphabet)
```

Comment:
According to comments in the test this doesn't work right:
```cpp
// letter3 = dna4::T; does not work
```

Instead of having assignment / construction from the variant_type, why not allow assignment / construction from the actual alphabet types, like e.g.:
```cpp
template <typename type>
    requires meta::in<meta::list<first_alphabet_type, alphabet_types...>, type>::value
constexpr union_alphabet(type const & alphabet) :
{
    _value = prefix_sum_sizes[meta::find_index<meta::list<first_alphabet_type, alphabet_types...>, type>::value] + static_cast<rank_type>(to_rank(alphabet));
}
```
This should also be faster and you can probably get rid of the variant altogether? But maybe I am missing something?